docs: running cockroachdb on kubernetes locally

jseldess opened the pull request:
This PR adds a new page explaining how to test out CockroachDB on Kubernetes using `minikube` and the alpha Pet Set feature. It's based on @tschottdorf's and @a-robinson's  CockroachDB example files in the [Kubernetes repo](https://github.com/kubernetes/kubernetes/tree/master/examples/cockroachdb). 

This still needs quite a bit of work, but I want to get some initial feedback.

Fixes #506

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/docs/584)
<!-- Reviewable:end -->

jseldess commented:
HTML version: http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/090f66b5dfe265cb8b811b30f686bea2d82ee390/orchestrate-cockroachdb-with-kubernetes.html

tschottdorf commented on orchestrate-cockroachdb-with-kubernetes.md:
```diff
@@ -0,0 +1,341 @@
+toc: false
+---
+
+[Kubernetes](http://kubernetes.io/) is an open-source system for automating the deployment, scaling, and management of containerized applications. Although Kubernetes is not yet production-ready for stateful applications such as CockroachDB, you can use Kubernete's [`minikube`](https://github.com/kubernetes/minikube) tool to test out CockroachDB on Kubernetes locally, using persistent volumes and ...
```
Kubernetes'

tschottdorf replied:
Below as well.

sploiselle commented:
Didn't test yet, but <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with a few really minor nits.

---

Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 231 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQ5v30SynerukzZRVNe:-KQ5v30SynerukzZRVNf:b-6swz0v) ([raw file](https://github.com/cockroachdb/docs/blob/090f66b5dfe265cb8b811b30f686bea2d82ee390/orchestrate-cockroachdb-with-kubernetes.md#L231)):*
> ```Markdown
> ~~~
> 
> {{site.data.alerts.callout_success}}The Pet Set configuration set all CockroachDB nodes to write to <code>stderr</code>, so if you ever need access to a pod/node's logs to troubleshoot, use <code>kubectl logs PODNAME</code> rather than checking the log on the pod itself.{{site.data.alerts.end}}  
> ```

"set all CockroachDB nodes" -> "sets all CockroachDB nodes"

---

*[_data/sidebar_doc.yml, line 246 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQ5x2TcTZDw_KuszZol:-KQ5x2TcTZDw_KuszZom:b-4xijn7) ([raw file](https://github.com/cockroachdb/docs/blob/090f66b5dfe265cb8b811b30f686bea2d82ee390/_data/sidebar_doc.yml#L246)):*
> ```YAML
>               thirdlevelitems:
> 
>                 - title: With Kubernetes
> ```

Think simply "Kubernetes" would be a marginal improvement.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/584#-:-KQ5xZbs23cF0-f1C3ns:b-6y38cz)*
<!-- Sent from Reviewable.io -->

jseldess commented:
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 7 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQ3sxPiBievcNWhIbei-r1-7:-KQ6974xF1g5bwFUuaK6:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/090f66b5dfe265cb8b811b30f686bea2d82ee390/orchestrate-cockroachdb-with-kubernetes.md#L7)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Below as well.
</details>
Done.

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 231 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQ5v30SynerukzZRVNe:-KQ68zgQypNZsKpMkqw1:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/090f66b5dfe265cb8b811b30f686bea2d82ee390/orchestrate-cockroachdb-with-kubernetes.md#L231)):*
<details><summary><i>Previously, sploiselle (Sean Loiselle) wrote…</i></summary>
> "set all CockroachDB nodes" -> "sets all CockroachDB nodes"
</details>
Done.

---

*[_data/sidebar_doc.yml, line 246 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQ5x2TcTZDw_KuszZol:-KQ68oeebrV7dzWzhDre:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/090f66b5dfe265cb8b811b30f686bea2d82ee390/_data/sidebar_doc.yml#L246)):*
<details><summary><i>Previously, sploiselle (Sean Loiselle) wrote…</i></summary>
> Think simply "Kubernetes" would be a marginal improvement.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/584)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
This reads quite well to someone who has no kubernetes experience (me). I did not actually run it, but it is completely coherent, and there were links to Kubernetes for every concept that I had questions about.

---

Review status: 0 of 2 files reviewed at latest revision, 5 unresolved discussions.

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 84 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQ6LJwNNFMYuwXe7dvt:-KQ6LJwNNFMYuwXe7dvu:b-7bvkj1) ([raw file](https://github.com/cockroachdb/docs/blob/dba2660c6f5fd78ebe91a9cfed332342269a77e8/orchestrate-cockroachdb-with-kubernetes.md#L84)):*
> ```Markdown
> In Kubernetes, pods are normally treated as stateless units. The [Pet Set](http://kubernetes.io/docs/user-guide/petset/) feature, however, causes pods to have persistent identity and storage. Using this feature, if a pod crashes or is otherwise removed, Kubernetes will create another pod with the same identity and storage (as demonstrated in [step 6](#step-6-simulate-node-failure)). 
> 
> Create the following Pet Set yaml file. Note that the `replicas: 5` line specifies that the CockroachDB cluster will contain five pods/nodes.
> ```

Maybe this is obvious to Kubernetes folks, but this was slightly unclear to me. I'm guessing this results in 5 pods with 1 node per pod, but it would be helpful if this was clarified.

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 210 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQ6O8087P0FvVwIXbUV:-KQ6O8087P0FvVwIXbUW:bpdrg9p) ([raw file](https://github.com/cockroachdb/docs/blob/dba2660c6f5fd78ebe91a9cfed332342269a77e8/orchestrate-cockroachdb-with-kubernetes.md#L210)):*
> ```Markdown
> ## Step 4. Start the CockroachDB cluster
> 
> To create pods and start the CockroachDB cluster, run the [`kubectl create`](http://kubernetes.io/docs/user-guide/kubectl/kubectl_create/) command against your Pet Set yaml file, in this case called `cockroachdb-petset.yaml`:
> ```

Can you also mention this filename above (as the header to the script itself)? It just makes things super easy when following along (if I was following along I would have already copied the bash script over and given it a random file name).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/584#-:-KQ6L2oOa1u0WI_shwtC:b-mtt2cy)*
<!-- Sent from Reviewable.io -->

tschottdorf commented on orchestrate-cockroachdb-with-kubernetes.md:
```diff
@@ -0,0 +1,341 @@
+
+## Step 2. Create persistent volumes and volume claims
+
+Each CockroachDB node will run in a [pod](http://kubernetes.io/docs/user-guide/pods/), which is a group of one or more Docker containers that can share resources. In our case, each pod will have only one container running a CockroachDB node. 
```
> Due to running as a PetSet, the storage the containers will use is a little more complicated because we want it to be persistent and owned by a specific identity of a Cockroach node, which necessitates the use of persistent volumes and associated claims. Use the below script as a black-box for setting up the correct volumes and claims to make this example work.

a-robinson replied:
Including something like this may be good, but if we do we should clarify what it "is a little more complicated" than

tschottdorf replied:
I'm not sure what the right balance between useful detail and overly distracting from the actual example is, but it likely depends on the target demographic. Maybe someone who doesn't know kubernetes well can weigh in.

a-robinson commented:
Looks great! Do you have anything in mind for how to squeeze in instructions for creating PVs when running in a cloud environment? Right now, the guide almost makes it sound like running locally is the only option

---

Review status: 0 of 2 files reviewed at latest revision, 14 unresolved discussions.

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 40 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQ7HP2aUZ2MEobklFys:-KQ7HP2aUZ2MEobklFyt:b14ai2c) ([raw file](https://github.com/cockroachdb/docs/blob/dba2660c6f5fd78ebe91a9cfed332342269a77e8/orchestrate-cockroachdb-with-kubernetes.md#L40)):*
> ```Markdown
>   name: pv${i}
>   labels:
>     type: local
> ```

Add the label `app: cockroachdb` to make deleting the PVs easier. You must have copied this from an older version of my PR before I added it.

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 67 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQ7HkyELgPE3i5d-jBF:-KQ7HkyELgPE3i5d-jBG:b-d6f0n6) ([raw file](https://github.com/cockroachdb/docs/blob/dba2660c6f5fd78ebe91a9cfed332342269a77e8/orchestrate-cockroachdb-with-kubernetes.md#L67)):*
> ```Markdown
> ~~~
> 
> Use the [`kubectl get pv`](http://kubernetes.io/docs/user-guide/kubectl/kubectl_get/) command to verify that persistent volumes and corresponding claims were created successfully:
> ```

s/that persistent volumes/that the persistent volumes/

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 82 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQ7HxY_4QxT3a-CLNnt:-KQ7HxY_4QxT3a-CLNnu:b-dfetzs) ([raw file](https://github.com/cockroachdb/docs/blob/dba2660c6f5fd78ebe91a9cfed332342269a77e8/orchestrate-cockroachdb-with-kubernetes.md#L82)):*
> ```Markdown
> ## Step 3. Create the Pet Set configuration
> 
> In Kubernetes, pods are normally treated as stateless units. The [Pet Set](http://kubernetes.io/docs/user-guide/petset/) feature, however, causes pods to have persistent identity and storage. Using this feature, if a pod crashes or is otherwise removed, Kubernetes will create another pod with the same identity and storage (as demonstrated in [step 6](#step-6-simulate-node-failure)). 
> ```

Optional, but I might be just a little more specific and replace "same identity and storage" with "same network identity and storage"

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 96 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQ7I9B6yc_rGoRJ1w3j:-KQ7I9B6yc_rGoRJ1w3k:b-236qtc) ([raw file](https://github.com/cockroachdb/docs/blob/dba2660c6f5fd78ebe91a9cfed332342269a77e8/orchestrate-cockroachdb-with-kubernetes.md#L96)):*
> ```Markdown
>   name: cockroachdb-public
>   labels:
>     app: cockroachdb-public
> ```

Again, this isn't quite the most recent version of the file, since I switched this to just be `app: cockroachdb` for easier deletion. I'd recommend re-copying the entire thing

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 277 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQ7IrD3N2U0IwAR4vvG:-KQ7IrD3N2U0IwAR4vvH:b3qx9p9) ([raw file](https://github.com/cockroachdb/docs/blob/dba2660c6f5fd78ebe91a9cfed332342269a77e8/orchestrate-cockroachdb-with-kubernetes.md#L277)):*
> ```Markdown
> Based on the `replicas: 5` line in the Pet Set configuration, Kubernetes ensure that five pods/nodes are running at all times. If a pod/node fails, Kubernetes will automatically create another pod/node with the same identity and storage.
> 
> To see this in action, kill one of the pods:
> ```

I might rephrase this to "kill one of the cockroachdb nodes", since the pod being restarted is more of an indirect consequence of killing the cockroachdb process

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 318 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQ7J9_PhCgPuDRVN_C7:-KQ7J9_PhCgPuDRVN_C8:b-1dfhtv) ([raw file](https://github.com/cockroachdb/docs/blob/dba2660c6f5fd78ebe91a9cfed332342269a77e8/orchestrate-cockroachdb-with-kubernetes.md#L318)):*
> ```Markdown
> cockroachdb-1   1/1       Running   0          19m
> cockroachdb-2   1/1       Running   0          19m
> cockroachdb-3   1/1       Running   4          11m
> ```

nit, but I might change the number of restarts from 4 to 1, just to match what users will see

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 338 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQ7JJOMCJsIlVerEKUj:-KQ7JJON0QdB3M6g9VSz:b-h32elf) ([raw file](https://github.com/cockroachdb/docs/blob/dba2660c6f5fd78ebe91a9cfed332342269a77e8/orchestrate-cockroachdb-with-kubernetes.md#L338)):*
> ```Markdown
> 
> ~~~ shell
> $ kubectl delete pods,pvc,petsets,svc -l app=cockroachdb
> ```

After grabbing the latest configs from my PR, this can just be:
kubectl delete petsets,pods,persistentvolumes,persistentvolumeclaims,services -l app=cockroachdb

(or the same command using the appropriate abbreviations, but spelling them out seemed more clear to new folks)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/584#-:-KQ7JXH8OOz5pGeEErH3:bh3sfof)*
<!-- Sent from Reviewable.io -->

jseldess commented:
Thanks, @a-robinson. Think I'll keep this PR focused on the local scenario and work on cloud deployment later. But I'll rework the intro to less misleading.

---

Review status: 0 of 2 files reviewed at latest revision, 14 unresolved discussions.

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 7 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQ3sxPiBievcNWhIbei-r2-7:-KQMNdrsr50vNYNAGi47:bmdrlqd) ([raw file](https://github.com/cockroachdb/docs/blob/dba2660c6f5fd78ebe91a9cfed332342269a77e8/orchestrate-cockroachdb-with-kubernetes.md#L7)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> >
> 
> Databases are stateful applications, and so running them on Kubernetes requires some of its more complex features, which are still being actively developed and thus only available in alpha versions of Kubernetes. There are easier ways to run CockroachDB for testing purposes on Kubernetes, but the method presented here is destined to become a production deployment once Kubernetes matures sufficiently.
> 
> >
> 
> CockroachDB will run as a PetSet - roughly speaking, a replication set whose instances are distinguishable and always bound back to the same persistent storage on restart.
</details>
Working on improving the intro, @tschottdorf. But can you provide more detail about some of the "easier ways to run CockroachDB" on Kubernetes?

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 40 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQ7HP2aUZ2MEobklFys:-KQMH8AA6LVsuKSocwUn:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/dba2660c6f5fd78ebe91a9cfed332342269a77e8/orchestrate-cockroachdb-with-kubernetes.md#L40)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary>
> Add the label `app: cockroachdb` to make deleting the PVs easier. You must have copied this from an older version of my PR before I added it.
</details>
Done.

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 67 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQ7HkyELgPE3i5d-jBF:-KQMFcpeYkkFVowX7Tnu:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/dba2660c6f5fd78ebe91a9cfed332342269a77e8/orchestrate-cockroachdb-with-kubernetes.md#L67)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary>
> s/that persistent volumes/that the persistent volumes/
</details>
Done.

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 82 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQ7HxY_4QxT3a-CLNnt:-KQMHArz77jL3B421gYO:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/dba2660c6f5fd78ebe91a9cfed332342269a77e8/orchestrate-cockroachdb-with-kubernetes.md#L82)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary>
> Optional, but I might be just a little more specific and replace "same identity and storage" with "same network identity and storage"
</details>
Done.

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 96 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQ7I9B6yc_rGoRJ1w3j:-KQMIaUX5rByMtGjUlkf:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/dba2660c6f5fd78ebe91a9cfed332342269a77e8/orchestrate-cockroachdb-with-kubernetes.md#L96)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary>
> Again, this isn't quite the most recent version of the file, since I switched this to just be `app: cockroachdb` for easier deletion. I'd recommend re-copying the entire thing
</details>
Done.

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 277 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQ7IrD3N2U0IwAR4vvG:-KQMIu_axvxsQ4M1IxLC:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/dba2660c6f5fd78ebe91a9cfed332342269a77e8/orchestrate-cockroachdb-with-kubernetes.md#L277)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary>
> I might rephrase this to "kill one of the cockroachdb nodes", since the pod being restarted is more of an indirect consequence of killing the cockroachdb process
</details>
Done.

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 318 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQ7J9_PhCgPuDRVN_C7:-KQMIxanCFgjtJJe6LEg:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/dba2660c6f5fd78ebe91a9cfed332342269a77e8/orchestrate-cockroachdb-with-kubernetes.md#L318)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary>
> nit, but I might change the number of restarts from 4 to 1, just to match what users will see
</details>
Done.

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 338 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQ7JJOMCJsIlVerEKUj:-KQMJ9GYNX68rKQvRaX3:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/dba2660c6f5fd78ebe91a9cfed332342269a77e8/orchestrate-cockroachdb-with-kubernetes.md#L338)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary>
> After grabbing the latest configs from my PR, this can just be:
> kubectl delete petsets,pods,persistentvolumes,persistentvolumeclaims,services -l app=cockroachdb
> 
> (or the same command using the appropriate abbreviations, but spelling them out seemed more clear to new folks)
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/584#-:-KQMNz3Ben-eyfm7IGdc:b-ld9zfn)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 7 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQ3sxPiBievcNWhIbei-r2-7:-KQMYZtQVmqywZ_fYN-b:bc45dgt) ([raw file](https://github.com/cockroachdb/docs/blob/dba2660c6f5fd78ebe91a9cfed332342269a77e8/orchestrate-cockroachdb-with-kubernetes.md#L7)):*
<details><summary><i>Previously, jseldess wrote…</i></summary>
> Working on improving the intro, @tschottdorf. But can you provide more detail about some of the "easier ways to run CockroachDB" on Kubernetes?
</details>
If you only want a single-node cluster, it's fairly simple to spin up a replication controller of size 1 that uses a persistent disk, but you won't be able to scale it up to more than one node. The config would look like https://github.com/cockroachdb/cockroach/issues/8876

It's even easier if you only want a single-node cluster and don't actually care about persistence (i.e. you're truly just kicking the tries):
kubectl run cockroachdb --image=cockroachdb/cockroach -- start

That'll give you a database that's just using the container's local filesystem, so if it ever gets killed, it'll lose its data. But if all you want is to play around a bit, it is technically running cockroach inside of kubernetes.

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 11 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQM_qS5H8PGW5DOtzdF:-KQM_qS5H8PGW5DOtzdG:bil42ox) ([raw file](https://github.com/cockroachdb/docs/blob/15143419dde0f375abebbd5826c1d56c885aa047/orchestrate-cockroachdb-with-kubernetes.md#L11)):*
> ```Markdown
> Databases are stateful applications, and so running them on Kubernetes requires some of its more complex features, which are still being actively developed and thus only available in alpha versions of Kubernetes. There are easier ways to run CockroachDB for testing purposes on Kubernetes, but the method presented here is destined to become a production deployment once Kubernetes matures sufficiently.
> 
> CockroachDB will run as a PetSet - roughly speaking, a replication set whose instances are distinguishable and always bound back to the same persistent storage on restart.
> ```

nit, but "always bound back" sounds funny to me. Perhaps use "bind" instead of "bound"?

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 30 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQM_lAeKbFC9DRpku_H:-KQM_lAeKbFC9DRpku_I:b51cn4e) ([raw file](https://github.com/cockroachdb/docs/blob/15143419dde0f375abebbd5826c1d56c885aa047/orchestrate-cockroachdb-with-kubernetes.md#L30)):*
> ```Markdown
> Each CockroachDB node will run in a [pod](http://kubernetes.io/docs/user-guide/pods/), which is a group of one or more Docker containers that can share resources. In our case, each pod will have only one container running a CockroachDB node. 
> 
> Normally, on-disk files in a pod are ephemeral. If a pod crashes and Kubernetes starts a new one, the on-disk files are lost. To prevent that from happening, you need to mount local temporary directories as [persistent volumes](http://kubernetes.io/docs/user-guide/persistent-volumes/) that will endure for as long as the Kubernetes cluster is running. You also need create [persistent volume claims](http://kubernetes.io/docs/user-guide/persistent-volumes/#persistentvolumeclaims); when pods are created later (one per node), each pod will request a persistent volume claim to "claim" durable storage for its node.
> ```

s/need create/need to create/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/584#-:-KQMaY_HeDzIW_wbaTlV:bnfp4nl)*
<!-- Sent from Reviewable.io -->

jseldess commented:
@a-robinson, @tschottdorf, please take one final look. I've reworked the intro and added a "Before You Begin" section to introduce Kubernetes concepts. 

---

Review status: 1 of 2 files reviewed at latest revision, 16 unresolved discussions.

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 7 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQ3sxPiBievcNWhIbei-r2-7:-KQNCrh1AcvG3ljtoATL:b-cocpd6) ([raw file](https://github.com/cockroachdb/docs/blob/dba2660c6f5fd78ebe91a9cfed332342269a77e8/orchestrate-cockroachdb-with-kubernetes.md#L7)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary>
> If you only want a single-node cluster, it's fairly simple to spin up a replication controller of size 1 that uses a persistent disk, but you won't be able to scale it up to more than one node. The config would look like https://github.com/cockroachdb/cockroach/issues/8876
> 
> It's even easier if you only want a single-node cluster and don't actually care about persistence (i.e. you're truly just kicking the tries):
> kubectl run cockroachdb --image=cockroachdb/cockroach -- start
> 
> That'll give you a database that's just using the container's local filesystem, so if it ever gets killed, it'll lose its data. But if all you want is to play around a bit, it is technically running cockroach inside of kubernetes.
</details>
Thanks, @a-robinson. This is helpful. For now, it seems best to just direct users to follow our approach, which is closer to what you want.

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 84 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQ6LJwNNFMYuwXe7dvt:-KQNHMKPKQ9pe4CuTO9Z:bsqxsmv) ([raw file](https://github.com/cockroachdb/docs/blob/dba2660c6f5fd78ebe91a9cfed332342269a77e8/orchestrate-cockroachdb-with-kubernetes.md#L84)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary>
> Maybe this is obvious to Kubernetes folks, but this was slightly unclear to me. I'm guessing this results in 5 pods with 1 node per pod, but it would be helpful if this was clarified.
</details>
I've tried to clarify this in the new "Before Your Begin" section and a few other places. Let me know if it's still confusing.

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 210 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQ6O8087P0FvVwIXbUV:-KQNHTSkfvnTaDynbC9Z:b-5xajg0) ([raw file](https://github.com/cockroachdb/docs/blob/dba2660c6f5fd78ebe91a9cfed332342269a77e8/orchestrate-cockroachdb-with-kubernetes.md#L210)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary>
> Can you also mention this filename above (as the header to the script itself)? It just makes things super easy when following along (if I was following along I would have already copied the bash script over and given it a random file name).
</details>
I've actually changed things to assume that the user will title the yaml file whatever.

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 11 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQM_qS5H8PGW5DOtzdF:-KQNHGC557x0hZV8YkPk:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/15143419dde0f375abebbd5826c1d56c885aa047/orchestrate-cockroachdb-with-kubernetes.md#L11)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary>
> nit, but "always bound back" sounds funny to me. Perhaps use "bind" instead of "bound"?
</details>
Done.

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 30 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQM_lAeKbFC9DRpku_H:-KQMducNeHidzliOMK_M:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/15143419dde0f375abebbd5826c1d56c885aa047/orchestrate-cockroachdb-with-kubernetes.md#L30)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary>
> s/need create/need to create/
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/584#-:-KQNHhkcSkoOaTK7q2tn:b-ejtph7)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 2 files at r2, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 18 unresolved discussions.

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 298 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQNIbJZ8HevDf78J9S0:-KQNIbJZ8HevDf78J9S1:bt7pf5n) ([raw file](https://github.com/cockroachdb/docs/blob/b447cf262deb925e39ae8ec8097b48f6976203b3/orchestrate-cockroachdb-with-kubernetes.md#L298)):*
> ```Markdown
> ## Step 3. Create the Pet Set and start the CockroachDB cluster
> 
> In Kubernetes, pods are normally treated as stateless units. The [Pet Set](http://kubernetes.io/docs/user-guide/petset/) feature, however, causes pods to have distinguisable network identity and persistent storage. Using this feature, if a pod crashes or is otherwise removed, Kubernetes will create another pod with the same network identity and storage (as demonstrated in [step 5](#step-5-simulate-node-failure)). 
> ```

typo in distinguishable

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 303 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQNIibCMam2k1YSN57p:-KQNIibCMam2k1YSN57q:bmb0gel) ([raw file](https://github.com/cockroachdb/docs/blob/b447cf262deb925e39ae8ec8097b48f6976203b3/orchestrate-cockroachdb-with-kubernetes.md#L303)):*
> ```Markdown
> 
> ~~~ yaml
> # Excerpted from # Excerpted from https://github.com/cockroachdb/cockroach/tree/develop/cloud/kubernetes/cockroachdb-petset.yaml
> ```

Doubled up

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/584#-:-KQNJ89S_d9PEUmxNudm:bnfp4nl)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 21 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQNL7VvjuOYJK8GAoPZ:-KQNL7VvjuOYJK8GAoP_:b-qmgfy1) ([raw file](https://github.com/cockroachdb/docs/blob/b447cf262deb925e39ae8ec8097b48f6976203b3/orchestrate-cockroachdb-with-kubernetes.md#L21)):*
> ```Markdown
> [minikube](http://kubernetes.io/docs/getting-started-guides/minikube/) | This is the tool you'll use to run a single-node Kubernetes cluster inside a VM on your laptop.
> [pod](http://kubernetes.io/docs/user-guide/pods/) | A pod is a group of one of more Docker containers. In our case, each pod will contain one Docker container running a single CockroachDB node. You'll start with 5 pods and grow to 6.
> [Pet Set](http://kubernetes.io/docs/user-guide/petset/) | You'll run CockroachDB as a Pet Set to ensure that each pod has distinguishable network identity and always binds back to the same persistent storage on restart. This feature is only available in alpha versions of Kubernetes. 
> ```

It'd be more correct to say something along the lines of "Pet Sets are only available in Kubernetes clusters that have alpha features enabled".  It's not a separate release track, it's just controlled by a flag in the main Kubernetes releases

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 23 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQNL_ihEN68QQa76L8I:-KQNL_ihEN68QQa76L8J:by1j669) ([raw file](https://github.com/cockroachdb/docs/blob/b447cf262deb925e39ae8ec8097b48f6976203b3/orchestrate-cockroachdb-with-kubernetes.md#L23)):*
> ```Markdown
> [Pet Set](http://kubernetes.io/docs/user-guide/petset/) | You'll run CockroachDB as a Pet Set to ensure that each pod has distinguishable network identity and always binds back to the same persistent storage on restart. This feature is only available in alpha versions of Kubernetes. 
> [persistent volume](http://kubernetes.io/docs/user-guide/persistent-volumes/) | Because each CockroachDB node must bind back to the same persistent storage on restart, you'll mount local temporary directories as persistent volumes that will endure for as long as the Kubernetes cluster is running.
> [persistent volume claim](http://kubernetes.io/docs/user-guide/persistent-volumes/#persistentvolumeclaims) | When pods are created (one per node), each pod will request a persistent volume claim to “claim” durable storage for its node.
> ```

I'd probably say "one per CockroachDB node", since otherwise this could lead people to think only one pod can be scheduled on each underlying VM (which Kubernetes calls "nodes" as well)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/584#-:-KQNMu9IZIjSklqGap98:bnfp4nl)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/584)*
<!-- Sent from Reviewable.io -->

jseldess commented:
Review status: 0 of 2 files reviewed at latest revision, 20 unresolved discussions.

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 21 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQNL7VvjuOYJK8GAoPZ:-KQOB4dBJOfDnaCCkCXo:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/b447cf262deb925e39ae8ec8097b48f6976203b3/orchestrate-cockroachdb-with-kubernetes.md#L21)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary>
> It'd be more correct to say something along the lines of "Pet Sets are only available in Kubernetes clusters that have alpha features enabled".  It's not a separate release track, it's just controlled by a flag in the main Kubernetes releases
</details>
Done.

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 23 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQNL_ihEN68QQa76L8I:-KQOBA0xLXPS4s4nRjRk:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/b447cf262deb925e39ae8ec8097b48f6976203b3/orchestrate-cockroachdb-with-kubernetes.md#L23)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary>
> I'd probably say "one per CockroachDB node", since otherwise this could lead people to think only one pod can be scheduled on each underlying VM (which Kubernetes calls "nodes" as well)
</details>
Done.

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 298 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQNIbJZ8HevDf78J9S0:-KQOBHlS--1GchFDyCQX:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/b447cf262deb925e39ae8ec8097b48f6976203b3/orchestrate-cockroachdb-with-kubernetes.md#L298)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> typo in distinguishable
</details>
Done.

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 303 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQNIibCMam2k1YSN57p:-KQOBLaEZQKauINBcyCL:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/b447cf262deb925e39ae8ec8097b48f6976203b3/orchestrate-cockroachdb-with-kubernetes.md#L303)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Doubled up
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/584)*
<!-- Sent from Reviewable.io -->

jseldess commented:
@sploiselle, would you mind testing out the steps before I merge?

sploiselle commented:
Review status: 0 of 2 files reviewed at latest revision, 24 unresolved discussions.

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 37 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQQpl0tJ1WVAPVSaSLm:-KQQpl0tJ1WVAPVSaSLn:b-4q6qsm) ([raw file](https://github.com/cockroachdb/docs/blob/b6f34de102c369180543b83240281a7a40975e8a/orchestrate-cockroachdb-with-kubernetes.md#L37)):*
> ```Markdown
> ## Step 2. Create persistent volumes and volume claims
> 
> To create persistent volumes and volume claims, run the following Bash script or individual `kubetcl` commands:
> ```

Storing a copy of this file as something like `kubernetes-config.sh` and hosting it would be great. This makes it easy to download and then run `sh kubernetes-config.sh`

This lets you convert this into steps and then expose the contents of the file with the accordion/on-page toggle control. Something like:

1. Download the `kubernetes-config.sh` script.
    [Script Details]
2. Run the script:
   `sh kubernetes-config.sh`

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 300 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQQqFVs-A-U0d8txuYO:-KQQqFVs-A-U0d8txuYP:b6q670c) ([raw file](https://github.com/cockroachdb/docs/blob/b6f34de102c369180543b83240281a7a40975e8a/orchestrate-cockroachdb-with-kubernetes.md#L300)):*
> ```Markdown
> In Kubernetes, pods are normally treated as stateless units. The [Pet Set](http://kubernetes.io/docs/user-guide/petset/) feature, however, causes pods to have distinguishable network identity and persistent storage. Using this feature, if a pod crashes or is otherwise removed, Kubernetes will create another pod with the same network identity and storage (as demonstrated in [step 5](#step-5-simulate-node-failure)). 
> 
> To create the Pet Set and start the CockrochDB cluster, create a yaml file with the contents below and then run `kubectl create -f <file-name.ymal>` against it. 
> ```

s/ymal/yaml

Making this downloadable, as well, would be great.

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 457 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQQru7UNpG9CF0Fm-9R:-KQQru7VM4iEa406Mvi2:be3j9u3) ([raw file](https://github.com/cockroachdb/docs/blob/b6f34de102c369180543b83240281a7a40975e8a/orchestrate-cockroachdb-with-kubernetes.md#L457)):*
> ```Markdown
> Then run some [CockroachDB SQL statements](sql-statements.html):
> 
> ~~~ shell
> ```

Mention the bug with the SQL shell for users' sanity.

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 485 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQQtBsxImWoQFGPb0h6:-KQQtBsxImWoQFGPb0h7:boliqk4) ([raw file](https://github.com/cockroachdb/docs/blob/b6f34de102c369180543b83240281a7a40975e8a/orchestrate-cockroachdb-with-kubernetes.md#L485)):*
> ```Markdown
> 
> ~~~ shell
> $ kubectl exec cockroachdb-3 -- /bin/bash -c "while true; do kill 1; done"
> ```

include output `error: error executing remote command: error executing command in container: Error executing in Docker Container: 137` to show that is the desired response and it isn't an *actual* error

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/584)*
<!-- Sent from Reviewable.io -->

jseldess commented:
Review status: 0 of 2 files reviewed at latest revision, 24 unresolved discussions.

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 37 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQQpl0tJ1WVAPVSaSLm:-KQRneu_uiWGUWBw7qAI:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/b6f34de102c369180543b83240281a7a40975e8a/orchestrate-cockroachdb-with-kubernetes.md#L37)):*
<details><summary><i>Previously, sploiselle (Sean Loiselle) wrote…</i></summary>
> Storing a copy of this file as something like `kubernetes-config.sh` and hosting it would be great. This makes it easy to download and then run `sh kubernetes-config.sh`
> 
> This lets you convert this into steps and then expose the contents of the file with the accordion/on-page toggle control. Something like:
> 
> 1. Download the `kubernetes-config.sh` script.
>     [Script Details]
> 2. Run the script:
>    `sh kubernetes-config.sh`
</details>
Done.

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 300 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQQqFVs-A-U0d8txuYO:-KQRndu3qriYx6vBpn7V:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/b6f34de102c369180543b83240281a7a40975e8a/orchestrate-cockroachdb-with-kubernetes.md#L300)):*
<details><summary><i>Previously, sploiselle (Sean Loiselle) wrote…</i></summary>
> s/ymal/yaml
> 
> Making this downloadable, as well, would be great.
</details>
Done.

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 457 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQQru7UNpG9CF0Fm-9R:-KQRzualX_xSFtY5IvPc:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/b6f34de102c369180543b83240281a7a40975e8a/orchestrate-cockroachdb-with-kubernetes.md#L457)):*
<details><summary><i>Previously, sploiselle (Sean Loiselle) wrote…</i></summary>
> Mention the bug with the SQL shell for users' sanity.
</details>
Done.

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 485 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQQtBsxImWoQFGPb0h6:-KQS-F80CM7Nm7JrWmb9:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/b6f34de102c369180543b83240281a7a40975e8a/orchestrate-cockroachdb-with-kubernetes.md#L485)):*
<details><summary><i>Previously, sploiselle (Sean Loiselle) wrote…</i></summary>
> include output `error: error executing remote command: error executing command in container: Error executing in Docker Container: 137` to show that is the desired response and it isn't an *actual* error
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/584)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 2 files at r5, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 25 unresolved discussions.

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 70 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQSDMnwcVxy1C93QbL2:-KQSDMnwcVxy1C93QbL3:b-lvtz71) ([raw file](https://github.com/cockroachdb/docs/blob/f465dbe2448965ea348409e46cd8f9aaf9ed8058/orchestrate-cockroachdb-with-kubernetes.md#L70)):*
> ```Markdown
>    ~~~
> 
> 4. Wait a bit and then verify that five pods were created successfully. If you don't see five pods, wait longer and try again.
> ```

For what it's worth, I initially misinterpreted "try again" here to mean "try creating the resources again" rather than "try running `kubectl get pods` again". Maybe it's just me, but maybe s/try again/check again/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/584#-:-KQSEcXdb0RDEX5xNt7h:bnfp4nl)*
<!-- Sent from Reviewable.io -->

jseldess commented:
Review status: 1 of 2 files reviewed at latest revision, 25 unresolved discussions.

---

*[orchestrate-cockroachdb-with-kubernetes.md, line 70 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/docs/584#-KQSDMnwcVxy1C93QbL2:-KQSF-kN2NCR2JDimsFr:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/f465dbe2448965ea348409e46cd8f9aaf9ed8058/orchestrate-cockroachdb-with-kubernetes.md#L70)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary>
> For what it's worth, I initially misinterpreted "try again" here to mean "try creating the resources again" rather than "try running `kubectl get pods` again". Maybe it's just me, but maybe s/try again/check again/
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/584)*
<!-- Sent from Reviewable.io -->
